algorithms/linfa-trees/src/decision_trees/algorithm.rs

```diff
@@ -196,13 +197,135 @@ impl<F: Float, L: Label + std::fmt::Debug> TreeNode<F, L> {
         }
     }
 
+    // Will find the best split of the possible splits available
+    // May consider only one randomly selected split if the hyperparameter `random_split` is enabled
+    fn best_split<D: Data<Elem = F>, T: AsTargets<Elem = L> + Labels<Elem = L>>(
+        parent_class_freq: &HashMap<L, f32>,
+        feature_idx: usize,
+        sorted_index: &SortedIndex<F>,
+        data: &DatasetBase<ArrayBase<D, Ix2>, T>,
+        mask: &RowMask,
+        hyperparameters: &DecisionTreeParams<F, L>,
+        target: ArrayBase<ndarray::ViewRepr<&L>, Ix1>,
+        best: &mut Option<(usize, F, f32)>,
+    ) {
+        let mut right_class_freq = parent_class_freq.clone();
+        let mut left_class_freq = HashMap::new();
+
+        // We keep a running total of the aggregate weight in the right split
+        // to avoid having to sum over the hash map
+        let total_weight = parent_class_freq.values().sum::<f32>();
+        let mut weight_on_right_side = total_weight;
+        let mut weight_on_left_side = 0.0;
+
+        // Consider a single random split rather than an optimal split
+        // if `random_split` is enabled.
+        let mut random_split_idx = 0;
+
+        if hyperparameters.random_split {
+            let valid_split_points: Vec<usize> = (0..mask.mask.len())
+                .filter(|&i| mask.mask[sorted_index.sorted_values[i].0])
+                .collect();
+            random_split_idx = rand::thread_rng().gen_range(0..valid_split_points.len());
+        }
```

Comment:
I assume the point here is to randomly select a valid point to split at, correct? If so, this doesn't seem to do that to me. The rng function [generates a value from the provided range](https://docs.rs/rand/0.8.4/rand/trait.Rng.html#method.gen_range), but this doesn't correspond to one of the valid mask values.

Also, if this is being computed anyway doesn't it make sense to just loop over the valid mask values instead of having a check inside the following for loop?

Reply:
Yeah, you're right. I thought I had switched it to `choose` at some point but it doesn't look like it saved. 

I thought of just looping throguh `valid_split_points`, but hesitated since it would add an additional unnecessary loop for the decision tree use case. But I guess the optimization matters a lot more in the ensemble scenario so I'll switch this.

Reply:
doesn't a normal decision tree just want the valid_split_points as well? It seems like it's doing a check inside the loop and continuing if it's invalid, unless I'm misunderstanding?

Reply:
Yeah, I was just originally trying to avoid the pass through all observations to filter and then going through the valid points when we could just go through all of the observations and skip the invalid ones in one pass as it was doing before. The valid points were originally only being computed for the extra trees case but now I'm doing it regardless bc it's definitely not worth worrying about the extra loop for one decision tree.

Reply:
Oh I see now. We can avoid changing the loop then, instead just fixing the random index selection